wallet-new/src/Cardano/Wallet/Kernel/Submission.hs

```diff
@@ -0,0 +1,393 @@
+
+-- | Overrides the 'Scheduler' with an input transaction and a
+-- custom 'Schedule'. Useful to force dispatching, especially in tests.
+unsafeSchedule :: WalletSubmission m
+               -> Slot
+               -> Scheduled
+               -> WalletSubmission m
+unsafeSchedule ws slot scheduled@(Scheduled (txId, txAux, _)) =
+    ws & over (wsState . wssPendingSet) (addOneToPending txId txAux)
+       . over (wsState . wssScheduler) overrideSchedule
+    where
+        overrideSchedule :: Scheduler -> Scheduler
+        overrideSchedule (Scheduler s) =
+            Scheduler (IntMap.insertWith mappend slot [scheduled] s)
+
+
+-- | A getter to the local pending set stored in this 'WalletSubmission'.
+localPendingSet :: Getter (WalletSubmission m) Pending
+localPendingSet = wsState . wssPendingSet
+
+getCurrentSlot :: Getter (WalletSubmission m) Slot
+getCurrentSlot = wsState . wssCurrentSlot
+
+getScheduler :: Getter (WalletSubmission m) Scheduler
+getScheduler = wsState . wssScheduler
+
+
+--
+-- Internal API
+--
+
+-- | Schedule the full list of pending transactions.
+-- The transactions will be scheduled immediately in the next 'Slot'.
+schedule :: WalletSubmission m
+         -> Pending
+         -> WalletSubmission m
+schedule ws pending =
+    let currentSlot = ws ^. wsState . wssCurrentSlot
+    in ws & over (wsState . wssScheduler) (scheduleNext currentSlot)
+    where
+        scheduleNext :: Slot -> Scheduler -> Scheduler
+        scheduleNext currentSlot (Scheduler s) =
+            let txs     = map toEntry (pending ^. pendingTransactions . fromDb . to M.toList)
+                toEntry (txId, txAux) = Scheduled (txId, txAux, Schedule 0)
+            in Scheduler (IntMap.insertWith mappend (succ currentSlot) txs s)
+
+-- | Purges all the scheduled transactions in the given 'Slot'. This is
+-- automatically called by 'tick', and it's responsibility of @rho@ to
+-- correctly reschedule all the transaction.
+purgeSlot :: Slot -> Scheduler -> Scheduler
+purgeSlot slot (Scheduler s) = Scheduler (IntMap.delete slot s)
+
+-- TODO(adn): Better placed into module 'Cardano.Wallet.Kernel.DB.Spec'?
+
+-- | Computes the union between two 'Pending' sets.
+unionPending :: Pending -> Pending -> Pending
+unionPending (Pending new) (Pending old) =
+    Pending (M.union <$> new <*> old)
+
+-- | Computes the difference between two 'Pending' sets.
+differencePending :: Pending -> Pending -> Pending
+differencePending (Pending new) (Pending old) =
+    Pending (M.difference <$> new <*> old)
+
+-- | Adds a single element to the 'Pending' set.
+addOneToPending :: Core.TxId -> Core.TxAux -> Pending -> Pending
+addOneToPending txId aux (Pending p) = Pending (M.insert txId aux <$> p)
+
+-- Ready-to-use 'ResubmissionFunction's.
+
+-- | A 'RetryPolicy' is simply a function which instruct the 'Scheduler' when
+-- to attempt resubmitting the given 'Scheduled' item.
+type RetryPolicy = Slot -> Slot
+
+--
+-- Stock retry policies inspired by the 'retry' package.
```

Comment:
Nice, I like this abstraction :)

Reply:
Do you think though that it might make sense to change it to `Slot -> Maybe Slot`, so that these retry policies can decide when to give up, rather than hardcoding the `255` limit in `defaultResubmissionFunction`?